third_party/move/move-compiler-v2/src/bytecode_generator.rs

```diff
@@ -1396,6 +1802,315 @@ impl<'env> Generator<'env> {
         let rhs_vars = rhs.free_vars_and_used_params(&param_symbols);
         lhs_vars.intersection(&rhs_vars).next().is_some()
     }
+
+    fn branch_to_exit_if_false(&mut self, id: NodeId, cond: TempIndex, exit_path: Label) {
+        let fall_through = self.new_label(id);
+        self.emit_with(id, |attr| {
+            Bytecode::Branch(attr, fall_through, exit_path, cond)
+        });
+        self.emit_with(id, |attr| Bytecode::Label(attr, fall_through))
+    }
+
+    fn create_scope(
+        &mut self,
+        vars: impl Iterator<Item = (NodeId, Symbol)>,
+        as_ref: bool,
+    ) -> BTreeMap<Symbol, TempIndex> {
+        let mut scope = BTreeMap::new();
+        for (id, sym) in vars {
+            let mut ty = self.get_node_type(id);
+            if as_ref {
+                ty = Type::Reference(ReferenceKind::Immutable, Box::new(ty))
+            }
+            let temp = self.new_temp(ty);
+            scope.insert(sym, temp);
+            self.local_names.insert(temp, sym);
+        }
+        scope
+    }
+
+    /// Returns true if a value of given type need to be bound to a temporary, even if unused,
+    /// so it can be properly processed. Types which are known to be droppable
+    /// return false here.
+    fn ty_requires_binding(&self, ty: &Type) -> bool {
+        !self
+            .env()
+            .type_abilities(ty, &[])
+            .has_ability(Ability::Drop)
+    }
+}
+
+// ======================================================================================
+// Pattern coverage analysis
+
+/// Pattern coverage analysis is complicated by that patterns are
+/// matched sequentially (in contrast to so-called 'best fit' pattern matching).
+/// Consider the sequence of patterns from a match, `S{R{p}, q1}` and `S{_, q2}`.
+/// `R{p}` is only matched under the condition that `q1` is also matched. On
+/// the other hand, any arbitrary other value for the first parameter of `S` depends
+/// on that `q2` matches as well. These dependencies between patterns make it
+/// hard to decompose the analysis of completeness into independent sub-problems,
+/// and to apply some kind of algebraic simplification on patterns (e.g. something
+/// like `S{_, q1 | q2}` is _not_ equivalent to the above two patterns).
+///
+/// Another challenge in pattern coverage analysis is to produce useful error messages
+/// which show the counterexamples of missed patterns, but only up to the shape as the user
+/// has inspected values via patterns. For example, if the user has written `R{S{_}, _}`
+/// the counter example should not contain irrelevant other parts of the value (as e.g.
+/// in `R{T{x:_}, S}`). Moreover, missing patterns which can never be matched should
```

Comment:
"missing patterns which can never be matched" is a bit confusing => if there is a missing pattern (there exists atleast one value being matched that does not fit any match arm), it is an error, but there is no dead code; but if one pattern represents the superset of the other and appears first, then the other arm can never be matched (and therefore we have dead code). Would be helpful to clarify this in comments.